pkg/providers/openapi3/schema.go

```diff
@@ -0,0 +1,176 @@
+package openapi3
+
+import (
+	"errors"
+	"fmt"
+	openapi "github.com/getkin/kin-openapi/openapi3"
+	"github.com/jexia/semaphore/pkg/specs"
+	"github.com/jexia/semaphore/pkg/specs/labels"
+	"github.com/jexia/semaphore/pkg/specs/types"
+)
+
+func newSchemas(docs swaggers) (specs.Schemas, error) {
+	var result = make(specs.Schemas)
+
+	for file, doc := range docs {
+		endpoints, err := scanPaths(doc.Paths)
+		if err != nil {
+			return nil, fmt.Errorf("failed to scan document paths '%s': %w", file, err)
+		}
+
+		for _, endpoints := range endpoints {
+			for name, object := range endpoints.objects() {
+				name := getCanonicalName(doc, name)
+				// empty schemaRef, skipping
+				if object == nil || object.ref == nil {
+					continue
+				}
+
+				tpl, err := newTemplate(object.ref.Value)
+
+				if err != nil {
+					return nil, fmt.Errorf("failed to build property '%s' from file '%s': %w", name, file, err)
+				}
+
+				prop := &specs.Property{
+					Template: tpl,
+					Name:     name,
+					Label:    labels.Optional,
+				}
+
+				result[prop.Name] = prop
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// newTemplate builds a Property from the given swagger schemaRef component.
+// the function returns a property without name, the name should be set in the top caller.
+func newTemplate(schema *openapi.Schema) (tpl specs.Template, err error) {
+	switch schema.Type {
+	// handle as Message
+	case "object":
+		tpl, err = message(schema)
+		if err != nil {
+			return tpl, fmt.Errorf("failed to create object: %w", err)
+		}
+
+	// handle as Repeated
+	case "array":
+		tpl, err = repeated(schema)
+		if err != nil {
+			return tpl, fmt.Errorf("failed to create array: %w", err)
+		}
+
+	// check for OneOf, AnyOf, AllOf, Not
+	// and handle according to its type
+	case "":
+		switch {
+		case schema.OneOf != nil:
+			tpl, err = oneOf(schema)
+			if err != nil {
+				return tpl, fmt.Errorf("failed to create oneOf type: %w", err)
+			}
+		default:
+			return tpl, fmt.Errorf("unsupported mixed type")
+		}
+
+	// handle as scalar: string, integer, etc
+	default:
+		tpl, err = scalar(schema)
+		if err != nil {
+			return tpl, fmt.Errorf("failed to create scalar: %w", err)
+		}
+	}
+
+	return
```

Comment:
It is probably best to define `tpl, nil` over here to avoid confusion and hiding of logic.